swift3/iam.py

```diff
@@ -0,0 +1,436 @@
+
+    def check_condition(self, sid, req):
+        """
+        Check the conditions from the statement are satisfied.
+
+        :param sid: the ID of the statement defining the condition
+        :param req: the current request
+        """
+        for num, statement in enumerate(self._rules['Statement']):
+            # Statement ID is optional
+            cur_sid = statement.get('Sid', 'statement-id-%d' % num)
+            if sid != cur_sid:
+                continue
+            cond = statement.get('Condition', {})
+            for fname, operands in cond.items():
+                func = IamConditionFunc.get(fname)
+                for cond_key, values in operands.items():
+                    if cond_key not in IamConditionKey:
+                        self.logger.info(
+                            "IAM: ignoring condition %s: not implemented",
+                            cond_key)
+                        continue
+                    cond_val = self.resolve_cond_key(cond_key, req)
+                    if not func(cond_val, values):
+                        # One of the conditions is not satisfied
+                        return False
+            # All conditions are satisfied
+            return True
+        raise IAMException("BUG: statement %s not found" % sid)
+
+    def do_explicit_check(self, effect, action, req_res, req):
+        """
+        Lookup for an explicit deny or an explicit allow in the set of rules.
+
+        :param effect: one of RE_ALLOW or RE_DENY
+        :param req_res: the resource specified by the request
+        :returns: a tuple with a boolean telling of the rule has been matched
+            and the ID of the statement that matched.
+        """
+        for num, statement in enumerate(self._rules['Statement']):
+            # Statement ID is optional
+            sid = statement.get('Sid', 'statement-id-%d' % num)
+            self.logger.info("===> Checking statement %s (%s)",
+                             sid, statement['Effect'])
+            if statement['Effect'] != effect:
+                continue
+
+            # Check Action
+            for rule_action in statement['Action']:
+                if rule_action == action:
+                    # Found an exact match
+                    break
+                elif rule_action.endswith('*'):
+                    action_prefix = rule_action[:-1]
+                    if action.startswith(action_prefix):
+                        # Found a wildcard match
+                        break
+            else:
+                self.logger.info('Skipping %s, action %s is not in the list',
+                                 sid, action)
+                continue
+
+            for resource_str in statement['Resource']:
+                rule_res = IamResource(resource_str)
+
+                # check WildCard before everything else
+                if (rule_res.arn == ARN_WILDCARD_EVERYTHING and
+                        self.check_condition(sid, req)):
```

Comment:
We should pass the `statement` directly to avoid going through all the statements again. (current complexity: O(n^2))

Reply:
Ok, I can fix that.